graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/rest/PipelineResource.java

```diff
@@ -250,6 +272,111 @@ public PipelineSource update(@ApiParam(name = "id") @PathParam("id") String id,
         return PipelineSource.fromDao(pipelineRuleParser, savedPipeline);
     }
 
+    public static record RoutingRequest(
+            @JsonProperty(value = "input_id", required = true) String inputId,
+            @JsonProperty(value = "stream_id", required = true) String streamId,
+            @JsonProperty(value = "remove_from_default", required = true) boolean removeFromDefault
+    ) {}
+
+    @ApiOperation(value = "Add a stream routing rule to the default routing pipeline.")
+    @Path("/routing")
+    @PUT
+    @AuditEvent(type = PipelineProcessorAuditEventTypes.PIPELINE_UPDATE)
+    public PipelineSource routing(@ApiParam(name = "body", required = true) @NotNull RoutingRequest request) throws NotFoundException {
+        checkPermission(RestPermissions.STREAMS_EDIT, request.streamId());
+        checkPermission(PipelineRestPermissions.PIPELINE_RULE_CREATE);
+
+        String streamName;
+        try {
+            final Stream stream = streamService.load(request.streamId());
+            streamName = stream.getTitle();
+        } catch (NotFoundException e) {
+            throw new InternalServerErrorException(f("Unable to load stream %s", request.streamId()), e);
+        }
+
+        RuleDao ruleDao = createRoutingRule(request, streamName);
+        PipelineDao pipelineDao;
+        try {
+            pipelineDao = pipelineService.loadByName(GL_INPUT_ROUTING_PIPELINE);
+        } catch (NotFoundException e) {
+            // Create pipeline with first rule
+            return createRoutingPipeline(ruleDao);
+        }
+
+        // Add rule to existing pipeline
+        PipelineSource pipelineSource = PipelineSource.fromDao(pipelineRuleParser, pipelineDao);
+        final List<String> rules0 = pipelineSource.stages().get(0).rules();
+        if (rules0.stream().filter(ruleRef -> ruleRef.equals(ruleDao.title())).findFirst().isEmpty()) {
+            rules0.add(ruleDao.title());
+            PipelineSource updated = pipelineSource.toBuilder()
```

Comment:
we need to return the updated pipeline here. Maybe change the code like that:
```
            pipelineSource = pipelineSource.toBuilder()
                    .source(createPipelineString(pipelineSource))
                    .build();
            update(pipelineDao.id(), pipelineSource);
```

Reply:
fixed